File: Queue.js
function Queue(){
	this.dataStore = [];
	this.head = 0;
	this.tail = 0;
	this.enqueue = enqueue;
	this.dequeue = dequeue;
	this.front = front;
	this.back = back;
	this.empty = empty;
	this.length = 0;
	this.toString = toString;
}
function enqueue(element){
	this.length++;
	this.dataStore[this.tail++] = element;
}
function dequeue(){
	if(!this.empty()) return this.dataStore[this.head++];
	else return false;
}
function front(){
	if(!this.empty()) return this.dataStore[this.head];
	else return false;
}
function back(){
	if(!this.empty()) return this.dataStore[this.tail-1];
	else return false;	
}
function empty(){
	if(this.length === 0) return true;
	else return false;
}
function toString(){
	if(!this.empty()){
		var result = "";
		for(var i=this.head; i<this.tail; i++){
			result += this.dataStore[i] + "\n";
		}
		return result;
	} else {
		return false;
	}
}

File: Queue.java
class Queue<Item>{
	
	class Node<Item>{
		Node next;
		Node prev;
		Item key;
		public Node(Item key){
			this.key = key;
		}
	}
	
	Node head;
	Node tail;
	int length;
	
	public Queue(){
		head = null;
		tail = null;
		length = 0;
	}
	
	public Item front(){
		return (Item) head.key;
	}
	
	public Item back(){
		return (Item) tail.key;
	}
	
	public boolean isEmpty(){
		return head == null;
	}
	
	public Item enqueue(Item key){
		Node n = new Node(key);
		if(head == null){
			head = n;
			tail = head;
		} else {
			n.prev = tail.prev;
			tail.next = n;
			tail = n;
		}
		return (Item) n.key;
	}
	
	public Item dequeue(){
		if(!isEmpty()){
			Item temp = (Item) head.key;
			head = head.next;
//			head.next.prev = head;
			length--;
			return temp;
		}
		return null;
	}
	
	public void print(){
		Node n = this.head;
		n = n.next;
		while(n != tail){
			System.out.print(n.key + ", ");
			n = n.next;
		}
		System.out.print("\n");
	}
		
	public static void main(String[] args) {
		Queue<Integer> q = new Queue<Integer>();
		q.enqueue(3);
		q.enqueue(21);
		q.enqueue(15);
		q.enqueue(0);
		q.enqueue(101);
		q.enqueue(8);
		
		q.print();
		q.dequeue();
		q.print();
	}
}